dynawo/sources/Solvers/FixedTimeStep/SolverTRAP/DYNSolverTRAP.cpp

```diff
@@ -0,0 +1,207 @@
+// for power systems.
+//
+
+/**
+ * @file  DYNSolverTRAP.cpp
+ *
+ * @brief Trapezoidal solver implementation
+ *
+ * Trapezoidal solver is based on Trapezoidal Method that is nearly the same as Back Euler one.
+ *
+ */
+
+#include "DYNSolverTRAP.h"
+
+#include <cmath>
+#include <iostream>
+#include <iomanip>
+#include <vector>
+
+#include <boost/shared_ptr.hpp>
+
+#include "PARParametersSet.h"
+#include "PARParameter.h"
+
+#include "DYNMacrosMessage.h"
+#include "DYNSolverKINEuler.h"
+#include "DYNSolverKINAlgRestoration.h"
+#include "DYNTrace.h"
+#include "DYNModel.h"
+
+using boost::shared_ptr;
+using std::endl;
+using std::max;
+using std::min;
+using std::ofstream;
+using std::stringstream;
+using std::set;
+using std::setw;
+using std::setfill;
+using std::vector;
+using std::map;
+using std::string;
+using std::make_pair;
+
+using parameters::ParametersSet;
+using timeline::Timeline;
+
+/**
+ * @brief SolverTRAPFactory getter
+ * @return A pointer to a new instance of SolverTRAPFactory
+ */
+extern "C" DYN::SolverFactory* getFactory() {
+  return (new DYN::SolverTRAPFactory());
+}
+
+/**
+ * @brief SolverTRAPFactory destroy method
+ */
+extern "C" void deleteFactory(DYN::SolverFactory* factory) {
+  delete factory;
+}
+
+/**
+ * @brief SolverTRAP getter
+ * @return A pointer to a new instance of SolverTRAP
+ */
+extern "C" DYN::Solver* DYN::SolverTRAPFactory::create() const {
+  DYN::Solver* solver(new DYN::SolverTRAP());
+  return solver;
+}
+
+/**
+ * @brief SolverTRAP destroy method
+ */
+extern "C" void DYN::SolverTRAPFactory::destroy(DYN::Solver* solver) const {
+  delete solver;
+}
+
+namespace DYN {
+
+std::string
+SolverTRAP::solverType() const {
+  return "TrapezoidalSolver";
+}
+
+void
+SolverTRAP::init(const shared_ptr<Model>& model, const double& t0, const double& tEnd) {
+  init_common(model, t0, tEnd);
+
+  solverKINYPrim_.reset(new SolverKINAlgRestoration());
+  solverKINYPrim_->init(model_, SolverKINAlgRestoration::KIN_DERIVATIVES);
+}
+
+void
+SolverTRAP::calculateIC() {
+  Trace::debug() << DYNLog(CalculateIC) << Trace::endline;
+  // Root evaluation before the initialization
+  // --------------------------------
+  ySave_.assign(vectorY_.begin(), vectorY_.end());
+  velocitySave_.assign(vectorYp_.begin(), vectorYp_.end());
+
+  int counter = 0;
+  bool change = true;
+
+  // Updating discrete variable values and mode
+  model_->copyContinuousVariables(&vectorY_[0], &vectorYp_[0]);
+  model_->evalG(tSolve_, g0_);
+  evalZMode(g0_, g1_, tSolve_);
+  model_->rotateBuffers();
+  state_.reset();
+  model_->reinitMode();
+
+  solverKINAlgRestoration_->setupNewAlgebraicRestoration(fnormtolAlg_, initialaddtolAlg_, scsteptolAlg_, mxnewtstepAlg_, msbsetAlg_, mxiterAlg_,
+                                                         printflAlg_);
+  solverKINYPrim_->setupNewAlgebraicRestoration(fnormtolAlg_, initialaddtolAlg_, scsteptolAlg_, mxnewtstepAlg_, msbsetAlg_, mxiterAlg_, printflAlg_);
```

Comment:
You could create a common method for the first lines until solverKINAlgRestoration_->setupNewAlgebraicRestoration and add:
velocitySave_.assign(vectorYp_.begin(), vectorYp_.end());
solverKINYPrim_->setupNewAlgebraicRestoration(fnormtolAlg_, initialaddtolAlg_, scsteptolAlg_, mxnewtstepAlg_, msbsetAlg_, mxiterAlg_, printflAlg_);
after it